ipld/retriever.go

```diff
@@ -45,59 +48,77 @@ func NewRetriever(dag format.DAGService) *Retriever {
 }
 
 // Retrieve retrieves all the data committed to DataAvailabilityHeader.
-// If not available locally, it aims to request from the network only 1/4 of the data and reconstructs other 3/4 parts
-// using the rsmt2d.Codec. It steadily tries to request other 3/4 data if 1/4 is not found within
-// RetrieveQuadrantTimeout or unrecoverable.
+//
+// If not available locally, it aims to request from the network only any 1/4 of the data(quadrant)
+// and reconstructs other 3/4 parts via rstm2d. If the requested quadrant is not available within
+// RetrieveQuadrantTimeout, it starts requesting more, one by one, until either the data is
+// reconstructed, context is canceled or ErrByzantine is generated.
 func (r *Retriever) Retrieve(ctx context.Context, dah *da.DataAvailabilityHeader) (*rsmt2d.ExtendedDataSquare, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel() // cancels all the ongoing requests if reconstruction succeeds early
+
 	log.Debugw("retrieving data square", "data_hash", hex.EncodeToString(dah.Hash()), "size", len(dah.RowsRoots))
 	ses, err := r.newSession(ctx, dah)
 	if err != nil {
 		return nil, err
 	}
-
-	for _, q := range newQuadrants(dah) {
-		eds, err := ses.retrieve(ctx, q)
-		if err == nil {
-			return eds, nil
-		}
-
-		var errByz *rsmt2d.ErrByzantineData
-		if errors.As(err, &errByz) {
-			return nil, NewErrByzantine(ctx, r.dag, dah, errByz)
+	defer ses.Close()
+
+	// wait for a signal to start reconstruction
+	// try until either success or context or bad data
+	for {
+		select {
+		case <-ses.Done():
+			eds, err := ses.Reconstruct()
+			if err == nil {
+				return eds, nil
+			}
+			// it might be a catastrophic error case, check for it
+			var errByz *rsmt2d.ErrByzantineData
+			if errors.As(err, &errByz) {
+				return nil, NewErrByzantine(ctx, r.dag, dah, errByz)
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
 		}
-		log.Warnw("not enough shares to reconstruct data square, requesting more...", "err", err)
-		// retry quadrants until we can reconstruct the EDS or error out
 	}
-
-	return nil, format.ErrNotFound
 }
 
-type retrieverSession struct {
+// retrievalSession represents a data square retrieval session
+// It manages the data square, quadrant request retries
+// and provides simple API to know when
+type retrievalSession struct {
 	dag   format.NodeGetter
 	adder *NmtNodeAdder
 
 	treeFn rsmt2d.TreeConstructorFn
 	codec  rsmt2d.Codec
 
 	dah            *da.DataAvailabilityHeader
-	squareLk       sync.RWMutex
-	square         [][]byte
 	squareImported *rsmt2d.ExtendedDataSquare
+
+	quadrants []*quadrant
+	squareLk  sync.RWMutex
+	square    [][]byte
+	shares    uint32
```

Comment:
would recommed to call this sharesCounter or sth more concise but clearer than just shares.